patches/Reflector.patch

```diff
@@ -0,0 +1,23 @@
+@package jetbrains/phpstorm-stubs
+
+--- Reflection/Reflector.php
++++ Reflection/Reflector.php
+@@ -1,7 +1,5 @@
+ <?php
+ 
+-use JetBrains\PhpStorm\Deprecated;
+-
+ /**
+  * <b>Reflector</b> is an interface implemented by all
+  * exportable Reflection classes.
+@@ -15,9 +13,8 @@
+      *
+      * @link https://php.net/manual/en/reflector.export.php
+      * @return string|null
+-     * @removed 8.0
++     * @removed 7.4
+      */
+-    #[Deprecated(since: '7.4')]
+     public static function export();
```

Comment:
Shouldn't this be fixed directly in phpstorm-stubs?

Reply:
It is: https://github.com/JetBrains/phpstorm-stubs/commit/e8310b02436892908a71387e9f18957c7e787465

But the build failed with any recent version of phpstorm-stubs. Example : https://github.com/phpstan/phpstan-src/runs/3533422043
I think it's because phpstorm-stubs is now only compatible with PHP 8.
Were you aware about this issue?

Reply:
It should be fine - PHPStan uses an emulative PHP 8 lexer for parsing phpstorm-stubs. It's already using PHP 8-only syntax in the currently used version. So the fact that phpstorm-stubs cannot be updated right now probably means that they're using maybe PHP 8.1 syntax, or there's a bug in PHP-Parser. It's worth investigating what exactly is going on here and what's the failure.

Reply:
I just re-push the branch where I've tried to update phpstorm-stubs, that way you can have a better look of the problem

Reply:
I'd appreciate if you could take a look and debug it. It's a matter of reproducing the problem on your machine, you should get the line of the parse error easily.

Reply:
Here the error:
```
PhpParser\Error: Syntax error, unexpected '}', expecting T_VARIABLE on line 24 on unknown line
#0 /code/phpstan-src/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Ast/Locator.php(75): PHPStan\Parser\PhpParserDecorator->parse('<?php\n\nnamespac...')
#1 /code/phpstan-src/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Ast/Locator.php(54): PHPStan\BetterReflection\SourceLocator\Ast\Locator->findReflectionsOfType(Object(PHPStan\BetterReflection\Reflector\ClassReflector), Object(PHPStan\BetterReflection\SourceLocator\Located\InternalLocatedSource), Object(PHPStan\BetterReflection\Identifier\IdentifierType))
#2 /code/phpstan-src/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/AbstractSourceLocator.php(50): PHPStan\BetterReflection\SourceLocator\Ast\Locator->findReflection(Object(PHPStan\BetterReflection\Reflector\ClassReflector), Object(PHPStan\BetterReflection\SourceLocator\Located\InternalLocatedSource), Object(PHPStan\BetterReflection\Identifier\Identifier))
#3 /code/phpstan-src/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/AggregateSourceLocator.php(35): PHPStan\BetterReflection\SourceLocator\Type\AbstractSourceLocator->locateIdentifier(Object(PHPStan\BetterReflection\Reflector\ClassReflector), Object(PHPStan\BetterReflection\Identifier\Identifier))
#4 /code/phpstan-src/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/MemoizingSourceLocator.php(40): PHPStan\BetterReflection\SourceLocator\Type\AggregateSourceLocator->locateIdentifier(Object(PHPStan\BetterReflection\Reflector\ClassReflector), Object(PHPStan\BetterReflection\Identifier\Identifier))
#5 /code/phpstan-src/vendor/ondrejmirtes/better-reflection/src/Reflector/ClassReflector.php(45): PHPStan\BetterReflection\SourceLocator\Type\MemoizingSourceLocator->locateIdentifier(Object(PHPStan\BetterReflection\Reflector\ClassReflector), Object(PHPStan\BetterReflection\Identifier\Identifier))
#6 /code/phpstan-src/src/Reflection/BetterReflection/BetterReflectionProvider.php(136): PHPStan\BetterReflection\Reflector\ClassReflector->reflect('Serializable')
#7 /code/phpstan-src/src/Rules/ClassCaseSensitivityCheck.php(30): PHPStan\Reflection\BetterReflection\BetterReflectionProvider->hasClass('Serializable')
#8 /code/phpstan-src/src/Rules/Classes/ExistingClassesInClassImplementsRule.php(41): PHPStan\Rules\ClassCaseSensitivityCheck->checkClassNames(Array)
#9 /code/phpstan-src/src/Analyser/FileAnalyser.php(82): PHPStan\Rules\Classes\ExistingClassesInClassImplementsRule->processNode(Object(PhpParser\Node\Stmt\Class_), Object(PHPStan\Analyser\MutatingScope))
#10 /code/phpstan-src/src/Analyser/NodeScopeResolver.php(398): PHPStan\Analyser\FileAnalyser->PHPStan\Analyser\{closure}(Object(PhpParser\Node\Stmt\Class_), Object(PHPStan\Analyser\MutatingScope))
#11 /code/phpstan-src/src/Analyser/NodeScopeResolver.php(257): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Class_), Object(PHPStan\Analyser\MutatingScope), Object(Closure))
#12 /code/phpstan-src/src/Analyser/FileAnalyser.php(190): PHPStan\Analyser\NodeScopeResolver->processNodes(Array, Object(PHPStan\Analyser\MutatingScope), Object(Closure))
#13 /code/phpstan-src/src/PhpDoc/StubValidator.php(108): PHPStan\Analyser\FileAnalyser->analyseFile('/code/phpstan-s...', Array, Object(PHPStan\Rules\Registry), Object(Closure))
#14 /code/phpstan-src/src/Command/AnalyseApplication.php(72): PHPStan\PhpDoc\StubValidator->validate(Array, true)
#15 /code/phpstan-src/src/Command/AnalyseCommand.php(229): PHPStan\Command\AnalyseApplication->analyse(Array, true, Object(PHPStan\Command\Symfony\SymfonyOutput), Object(PHPStan\Command\Symfony\SymfonyOutput), false, true, '/code/phpstan-s...', Array, Object(Symfony\Component\Console\Input\ArgvInput))
#16 /code/phpstan-src/vendor/symfony/console/Command/Command.php(255): PHPStan\Command\AnalyseCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /code/phpstan-src/vendor/symfony/console/Application.php(1009): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /code/phpstan-src/vendor/symfony/console/Application.php(273): Symfony\Component\Console\Application->doRunCommand(Object(PHPStan\Command\AnalyseCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#19 /code/phpstan-src/vendor/symfony/console/Application.php(149): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#20 /code/phpstan-src/bin/phpstan(102): Symfony\Component\Console\Application->run()
#21 /code/phpstan-src/bin/phpstan(103): {closure}()
#22 {main}
```

I'm not sure how it works but it seems that the code below has been extracted from this file: https://github.com/JetBrains/phpstorm-stubs/blob/e8310b02436892908a71387e9f18957c7e787465/Core/Core_c.php#L166
```php
<?php

namespace {
    /**
     * Interface for customized serializing.
     * @link https://php.net/manual/en/class.serializable.php
     */
    interface Serializable
    {
        /**
         * String representation of object.
         * @link https://php.net/manual/en/serializable.serialize.php
         * @return string|null The string representation of the object or null
         * @throws Exception Returning other type than string or null
         */
        public function serialize();
        /**
         * Constructs the object.
         * @link https://php.net/manual/en/serializable.unserialize.php
         * @param string $data The string representation of the object.
         * @return void
         */
        public function unserialize(#[LanguageLevelTypeAware(['8.0' => 'string'], default: '')] $data);
    }
}
```

So the problem is the annotation interpreted as a comment in the middle of the line resulting in a syntax error.

I reproduce this error with PHP-7.4 but not with PHP-8.0.

Reply:
For the record, I get the same error even after updating PHP-Parser.